### DISCUSSION TITLE
Good starting points to understand Silius better

### DISCUSSION THREAD

**Discussion by malm0d** [silius-rs/silius#316] [General]
I still consider myself pretty new to both Rust and Silius (and even AA) so I'm trying to better understand in more detail the architecture and codebase for Silius.

Whats are some good starting points to understand how: from when the UserOp is sent to the bundler url, how does it get handled and stored in the UO mempool, and how does the bundler identify those UOs and start bundling them?
(Basically steps 1, 2, and 3 from the diagram in `architecture/silius.png`)

Thanks! :)

**Comment by Vid201** [comment]
> I still consider myself pretty new to both Rust and Silius (and even AA) so I'm trying to better understand in more detail the architecture and codebase for Silius.
> 
> Whats are some good starting points to understand how: from when the UserOp is sent to the bundler url, how does it get handled and stored in the UO mempool, and how does the bundler identify those UOs and start bundling them? (Basically steps 1, 2, and 3 from the diagram in `architecture/silius.png`)
> 
> Thanks! :)

Hey, that's a good point, but something is definitely missing. I haven't had time to do a more detailed write-up (any help is appreciated). Here are some of the available things:
- https://hackmd.io/@Vid201/aa-bundler-rust
- http://silius.rs/
- https://github.com/silius-rs/silius/blob/main/docs/architecture/silius.png

But it doesn't go really into the details. It would be cool to add something like that to the Siolius website in the future.

 In short, there are three main components that communicate over gRPC - RPC server, user operation mempool, and bundling. RPC servers accept user operations and send them over to the user opinion mempool, where they are validated. The bundling component periodically (or manually) takes some user operations (sorted by gas) (over gRPC again), puts them into a bundle, and sends them over to the execution client. If P2P is enabled, then user operation mempool is connected to the P2P network as well, where it can receive and send additional user operations.
